### ISSUE TITLE
Strange case of invalid_params

### ISSUE THREAD

**Issue by maksimt** [bristolcrypto/SPDZ-2#31]
We are having an issue with SPDZ-2 (at least the latest commit, and the Oct 6 commit; haven't tested other commits yet) where we are seeing invalid_params exceptions and aren't sure where they're from.

Our workflow is as follows:

1. `Scripts/setup-online.sh 3 128 128` which calls `Fake-Offline`.

2. `compile.py` on a `.mpc` `script`

3. `Player-Online.x -pn 5000 -h 192.168.10.11 -lgp 128 -lg2 128 2 script` on each of 3 hosts (with different `player_id`s)

We're not sure where this is being thrown from because `invalid_params` occurs in multiple places, and doesn't provide details (a potential quick fix?):

```./.idea/workspace.xml:      <find>invalid_params</find>
./.idea/workspace.xml:      <find>'invalid_params'</find>
./Exceptions/Exceptions.h:class invalid_params:  public exception
./FHE/FFT_Data.cpp:	{ throw invalid_params(); }
./FHE/Matrix.cpp:  if ((n+1)!=A[0].size())  { throw invalid_params(); }
./FHE/NTL-Subs.cpp:  if (lg2pr==0) { throw invalid_params(); }
./FHE/NTL-Subs.cpp:    { throw invalid_params(); }
./FHE/NTL-Subs.cpp:                          throw invalid_params();
./FHE/NTL-Subs.cpp:             throw invalid_params();
./FHE/NTL-Subs.cpp:  if (n<2 || n>10) { throw invalid_params(); }
./FHE/NTL-Subs.cpp:        throw invalid_params();
./FHE/NTL-Subs.cpp:  if (lg2p0==0 || lg2p1==0) { throw invalid_params(); }
./FHE/NTL-Subs.cpp:    { throw invalid_params(); }
./Math/gf2n.cpp:  if (j==-1) { throw invalid_params(); }
./Math/gf2n.cpp:  if (2*(n-1)-64+t1>=64) { throw invalid_params(); }
./Math/gf2nlong.cpp:  if (j==-1) { throw invalid_params(); }
./Processor/Data_Files.cpp:      throw invalid_params();
./Processor/Instruction.cpp:            throw invalid_params();
```

**Comment by mkskeller** [comment]
Running it with valgrind or gdb will give you a stacktrace, so you can find out where exactly the exception is triggered.

**Comment by maksimt** [comment]
Thank you for the advice, you may be able to tell that I am new to C++. :)

```cpp
#6  0x000000000041930e in gf2n_short::init_field (nn=<optimized out>) at Math/gf2n.cpp:85
#7  0x0000000000413bb1 in read_setup (dir_prefix="Player-Data/3-128-128/") at Math/Setup.cpp:153
```

Unfortunately I am not able to understand what `gf2n_short::init_field` is trying to do or why it's throwing this error. I will try to see if I can recreate it following the tutorial script.

If it helps, the function is called with `nn=128` and the content of `Player-Data/3-128-128/Params-Data` is:
```
172035116406933162231178957667602464769
128
```

**Comment by rdragos** [comment]
I think you need to compile with USE_GF2N_LONG = 1 in CONFIG file.

**Comment by maksimt** [comment]
I set this up on `ubuntu 16.04.3`, `ntl-9.10.10`, `mpir-3.0.0`, `libsodium-1.0.8` (apt-get default version), `m4 1.4.17` (apt-get default version), `build-essentials` (apt-get default version) and the latest `SimpleOT` from pascholl's repo. If you think it's useful I can share my Vagrant/Puppet files that reproduce my environment (or your suggested canonical environment) automatically in a VM. `CONFIG.mine` is:
```
PREP_DIR='-DPREP_DIR="/dev/shm/Player-Data/"'
USE_NTL=1
MOD = -DMAX_MOD_SZ=6
```

Here is my process trying to get the tutorial to run:

1. If I compile *without* `-DINSECURE` I get the following error when running `setup-online.sh`:
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  You are trying to use insecure benchmarking functionality for preprocessing.
You can activate this at compile time by adding -DINSECURE to the compiler options.
Make sure to run make clean as well.
/home/ubuntu/srcs/SPDZ-2/Scripts/setup-online.sh: line 33: 17994 Aborted                 (core dumped) $SPDZROOT/Fake-Offline.x ${players} -lgp ${bits} -lg2 ${g} --default ${default}
dd: failed to open 'Player-Data/Private-Input-0': No such file or directory
dd: failed to open 'Player-Data/Private-Input-1': No such file or directory
```
(Note that setup_online.sh doesn't respect -DPREP_DIR)

2. If I compile with `-DINSECURE` and `USE_GF2N_LONG=0` I am able to run the tutorial successfully. However my script ends with `invalid_params` . `Params-Data` is the same as before.

3. If I compile with `-DINSECURE` and `USE_GF2N_LONG=1` I am still able to run the tutorial successfully. Additionally, my script is able to progress past the `invalid_params` exception, but now it ends with a `MacCheck Failure`:
```
Using indirect communication.
Opening file Programs/Public-Input/mlbox_sum_vector_3_100
Opening file /dev/shm/Player-Data/Private-Input-2
Opening file /dev/shm/Player-Data/Public-Output-2
Opening file /dev/shm/Player-Data/Private-Output-2
Locking for sync of thread 0
REWINDING - ONLY FOR BENCHMARKING
terminate called after throwing an instance of 'mac_fail'
  what():  MacCheck Failure
```

It looks like each host has different MAC keys `cat `ls Player* | sort`:
```
3 -81250840014629407012226663632514081298 05ab9bb7ccf6f27b341250ee6daf24643 24023987519428501276182903522140001896 7046b5187fa76c5fcfb49b0cb1a3b2ec3 64895093712566099364020509169679345665 8a1fe4d49aa2fdbf4b1637b06aabc762
```

```
3 58545420586832417831125381446342152468 718a2a055a9fdb8a0e7c1ab6868c6f9f3 -37869637118175854029283207101891817527 6d6597d07f961b83b74648844f7491f63 -49113160730274931139542334041261580567 13f2076658c961bf2c58320a88215a0
```

```
3 -41634341075728356651893346581264026919 27e8df131b27510fea6a34ebb607f7f73 -50334336542420144842031519535221073607 c33fedbe75775b86557769bf9882c76b3 -65421220818567511712630070470330260088 4bdb858e6d5f0ec716b72df7fa974356
```

-----

I tried running with `spdz2-offline.x` but also got an error:
```
~/srcs/SPDZ-2/spdz2-offline.x -p 0 -N 3 -h 192.168.10.11
ServerSocket is bound on port 4999
Waiting for player 0
Sent 0 to 192.168.10.11:4999
Connected to player 0
Waiting for player 1
Connected to player 1
Waiting for player 2
Connected to player 2
Player 0 started.
My Name = 192.168.10.11
My number = 0
Player 0 sent name (info only) 192.168.10.11
Client IP address: 192.168.10.11
Player 1 started.
Player 1 sent name (info only) 192.168.10.12
Client IP address: 192.168.10.12
Player 2 started.
Player 2 sent name (info only) 192.168.10.13
Client IP address: 192.168.10.13
3 players
Player 0 is running on machine 192.168.10.11
Player 1 is running on machine 192.168.10.12
Player 2 is running on machine 192.168.10.13
ServerSocket is bound on port 5000
Setting up send to self socket to 127.0.0.1:5000 with id 0x70
Setting up client to 192.168.10.12:5001 with id 0x73
Setting up send to self socket to 127.0.0.1:5000 with id 0x32
Setting up client to 192.168.10.12:5001 with id 0x35
Setting up client to 192.168.10.13:5002 with id 0x76
As a server, waiting for client with id 0x70 to connect.
Setting up parameters
m = 32768
Setting up client to 192.168.10.13:5002 with id 0x38
As a server, waiting for client with id 0x32 to connect.
Setting up parameters
m = 13175
	 p = 172035116406933162231178957667602464769  u = 1478524928  :   128 <= 128
log(slack): 0
Optimal p1 vs minimal: 200/199
increasing lengths: 323/201
increasing lengths: 324/202
Trying primes of length 324 and 202
m = 32768 too small, increasing it to 65536
	 p = 198766463529478683931867765928436695041  u = 1532887040  :   128 <= 128
log(slack): 0
Optimal p1 vs minimal: 202/201
increasing lengths: 324/203
increasing lengths: 325/204
Trying primes of length 325 and 204
Difference to suggestion for p0: 0, for p1: -1
p0 needs 6 words
p1 needs 4 words
Total length: 529
	 pr0 = 1898717602908035481601  :   71
	 pr1 = 60758963292782562508801  :   76
		Following should be both 1
		pr1 mod m = 1
		pr1 mod 2^lg2m = 1
terminate called after throwing an instance of 'std::runtime_error'
  what():  Compiled for GF(2^128) only. Change parameters or compile without USE_GF2N_LONG
Aborted (core dumped)
```

Meanwhile `ot-offline.x --output` seems to run.
```
Setting up parameters
m = 32768
	 p = 172035116406933162231178957667602464769  u = 1478524928  :   128 <= 128
my_num: 1
Got list of 3 players from file: 
    192.168.10.11
    192.168.10.12
    192.168.10.13
ServerSocket is bound on port 10001
Doing real base OTs
Setting up server with id 1
Setting up client to 192.168.10.13:10002 with id 5
		BaseTime: 0.816247
Setting up send to self socket to 127.0.0.1:10001 with id 0x4
Setting up client to 192.168.10.13:10002 with id 0x7
As a server, waiting for client with id 0x3 to connect.
As a server, waiting for client with id 0x4 to connect.
Setting up server with id 10
Set up with player 0 in thread 0 with id 10
Setting up client to 192.168.10.13:10002 with id 14
Set up with player 2 in thread 0 with id 14
Setup generators
Starting computation
Generated 1000 outputs
OT thread 0 finished
OT thread 1 finished
OT threads finished
thread 1 finished
Generator thread on average took time 0.170326
OTs on average took time 0.153887
Triple computation on average took time 0.000168663
Time: 0.160806
Throughput: 6218.67
Writing MAC key to /dev/shm/Player-Data/3-128-128/Player-MAC-Keys-P0
```
 The MAC keys are still different on each host and my script still ends with `MacCheck Failure`.

**Comment by rdragos** [comment]
It might be that you have to run a `make clean` command before `make` with `USE_GF2N_LONG = 1` flag set on.

**Comment by maksimt** [comment]
I'm compiling SPDZ-2 after I already make SimpleOT in its subdirectory, so `make clean` shouldn't be necessary.

What I'm concerned about is that `ot-offline` only generates MAC keys. Is it supposed to be run *after* and in addition to `spdz2-offline`? Should `Scripts/setup-online.sh` be used at all here?

**Comment by mkskeller** [comment]
- We strongly recommend running `make clean` after changing compile-time parameters because `make` does not know which files are affected by such changes.
- `spdz2-offline.x` is not compatible with `USE_GF2N_LONG = 1`.
- `Scripts/setup-online.sh` is a quick but insecure replacement for any offline phase, hence you need to use `-DINSECURE` in order to use it.
- If you use `Scripts/setup-online.sh` and would like to run the online phase on different hosts, you have to synchronize the `Player-Data` on all hosts.
- `mac_fail` is a catch-all failure (inherent to the protocol). We recommend re-running `Scripts/setup-online.sh` or the desired offline phase if it appears.
- Different MAC keys on different hosts is crucial to the security of the protocol.

**Comment by mkskeller** [comment]
Furthermore, note that you have to run `ot-offline.x -c` in order to use the output with the online phase.

**Comment by maksimt** [comment]
@mkskeller Than you for the advice. I wasn't able to run `ot-offline.x -c` (it throws `'invalid_commitment'`), but I was able to run it `-a`. This is what `PREP_DIR` looks like on the 3 hosts (with i in {0,1,2}):
```bash
├── [         42]  HOSTS
└── [         60]  Player-Data
    └── [        100]  3-128-128
        ├── [         44]  Params-Data
        ├── [         41]  Player-MAC-Keys-P{i}
        └── [      96000]  Triples-2-P{i}
```

After compiling our .mpc scripts and writing input `PREP_DIR` looks like this (note that documentation says to place input in `PREP_DIR/Player-Data/` while Processor.cpp:23 looks for it in `PREP_DIR`; I placed it in both locations):

```bash
├── HOSTS
├── input_{i}.bin
├── input_{i}.bin.out
├── Player-Data
│   ├── 3-128-128
│   │   ├── Params-Data
│   │   ├── Player-MAC-Keys-P{i}
│   │   └── Triples-2-P{i}
│   └── Private-Input-{i}
├── Private-Input-{i}
└── Programs
    ├── Bytecode
    │   ├── mlbox_normalized_sum_vector_3_100-0.bc
    │   ├── mlbox_sum_vector_3_100-0.bc
    │   └── mlbox_sum_vector_float_3_100-0.bc
    ├── Public-Input
    │   ├── mlbox_normalized_sum_vector_3_100
    │   ├── mlbox_sum_vector_3_100
    │   └── mlbox_sum_vector_float_3_100
    ├── Schedules
    │   ├── mlbox_normalized_sum_vector_3_100.sch
    │   ├── mlbox_sum_vector_3_100.sch
    │   └── mlbox_sum_vector_float_3_100.sch
    └── Source
        ├── mlbox_normalized_sum_vector_3_100.mpc
        ├── mlbox_sum_vector_3_100.mpc
        └── mlbox_sum_vector_float_3_100.mpc
```

**Is it a problem that in the Programs/Bytecode directory all hosts have -0.bc?**

Each input is a binary file containing a vector of integers that we are trying to sum.

At this point I start a server and call `Player-Online` on each host:

```
[...]/Player-Online.x -pn 5000 -h 192.168.10.11 -lgp 128 -lg2 128 {i} mlbox_sum_vector_3_100
```

The good news is that with your help we have made some progress; however now we are getting a new error:

```
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Sent 0 to 192.168.10.11:4999
My Name = 192.168.10.11
My number = 0
3 players
Player 0 is running on machine 192.168.10.11
Player 1 is running on machine 192.168.10.12
Player 2 is running on machine 192.168.10.13
ServerSocket is bound on port 5000
[New Thread 0x7ffff666b700 (LWP 4509)]
loading params from: /dev/shm/Player-Data/3-128-128/Params-Data
MAC Key p = 35572
MAC Key 2 = b09276de861470200b02bcd07a7446d5
Opening file Programs/Schedules/mlbox_sum_vector_3_100.sch
Number of threads I will run in parallel = 1
Number of program sequences I need to load = 1
Loading program 0 from Programs/Bytecode/mlbox_sum_vector_3_100-0.bc
mlbox_sum_vector_3_100-0 needs more secret gf2n_long memory, resizing to 8192
mlbox_sum_vector_3_100-0 needs more clear gf2n_long memory, resizing to 8192
mlbox_sum_vector_3_100-0 needs more secret gfp memory, resizing to 8592
mlbox_sum_vector_3_100-0 needs more clear gfp memory, resizing to 8192
mlbox_sum_vector_3_100-0 needs more clear integer memory, resizing to 8192
Cost of first tape:
  Type p
             0 =          0        Triples à           0
             0 =          0        Squares à           0
             0 =          0           Bits à           0
             0 =          0       Inverses à           0
             0 =          0     BitTriples à           0
             0 =          0 BitGF2NTriples à           0
  Type 2
             0 =          0        Triples à           0
             0 =          0        Squares à           0
             0 =          0           Bits à           0
             0 =          0       Inverses à           0
             0 =          0     BitTriples à           0
             0 =          0 BitGF2NTriples à           0
Total cost: 0
[New Thread 0x7ffff5e6a700 (LWP 4510)]
	I am in thread 0
Using single-threaded receiving
Setting up send to self socket to 127.0.0.1:5000 with id 0x0
Waiting for thread 0 to be ready
Setting up client to 192.168.10.12:5001 with id 0x3
Setting up client to 192.168.10.13:5002 with id 0x6
As a server, waiting for client with id 0x0 to connect.
	Set up player in thread 0
Setting up Data_Files in: /dev/shm/Player-Data/3-128-128/
done
Using indirect communication.
Opening file Programs/Public-Input/mlbox_sum_vector_3_100
Opening file /dev/shm/Player-Data/Private-Input-0
Opening file /dev/shm/Player-Data/Public-Output-0
Opening file /dev/shm/Player-Data/Private-Output-0
Locking for sync of thread 0
Sent 0 elements in 0 rounds
Purging preprocessed data because something is wrong
Setting up Data_Files in: /dev/shm/Player-Data/3-128-128/
done
Removing /dev/shm/Player-Data/3-128-128/Triples-2-P0
Removing /dev/shm/Player-Data/3-128-128/Triples-p-P0
Removing /dev/shm/Player-Data/3-128-128/Squares-2-P0
Removing /dev/shm/Player-Data/3-128-128/Squares-p-P0
Removing /dev/shm/Player-Data/3-128-128/Bits-2-P0
Removing /dev/shm/Player-Data/3-128-128/Bits-p-P0
Removing /dev/shm/Player-Data/3-128-128/Inverses-2-P0
Removing /dev/shm/Player-Data/3-128-128/Inverses-p-P0
Removing /dev/shm/Player-Data/3-128-128/BitTriples-2-P0
Removing /dev/shm/Player-Data/3-128-128/BitGF2NTriples-2-P0
Removing /dev/shm/Player-Data/3-128-128/Inputs-2-P0-0
Removing /dev/shm/Player-Data/3-128-128/Inputs-p-P0-0
Removing /dev/shm/Player-Data/3-128-128/Inputs-2-P0-1
Removing /dev/shm/Player-Data/3-128-128/Inputs-p-P0-1
Removing /dev/shm/Player-Data/3-128-128/Inputs-2-P0-2
Removing /dev/shm/Player-Data/3-128-128/Inputs-p-P0-2
terminate called after throwing an instance of 'file_error'
  what():  File Error : IO problem when buffering gfp from /dev/shm/Player-Data/3-128-128/Inputs-p-P0-0
```

Also, I apologize for asking many basic questions; my colleague who wrote the .mpc scripts and is more familiar with SPDZ-2 is on vacation.

**Comment by mkskeller** [comment]
- Would you mind filing a separate issue for `invalid_commitment` case?
- The `-0.bc` ending of files in `Programs/Bytecode` is as desired by design.
- In the `file_error` case, have you generated the offline data using `online_setup.sh` or `ot-offline.x`? The latter is incomplete and does not support inputs, for example, which you seem to use in your program.

**Comment by mkskeller** [comment]
Furthermore, where does the documentation say `PREP_DIR/Player-Data/`?

**Comment by mkskeller** [comment]
One more thought: How are you using `sgf2n` that you require `-lg2 128`? If you could settle for GF(2^40), many of the problems you've had wouldn't arise.
